bin/radical-pilot-agent-funcs

```diff
@@ -217,24 +254,50 @@ class Executor(object):
             try:
                 task = self._mpq_work.get(block=True, timeout=0.1)
 
-            except mp.Queue.Empty:
+            except queue.Empty:
                 continue
 
-            import pprint
-            pprint.pprint(task)
+          # import pprint
+          # pprint.pprint(task)
 
             tid   = task['uid']
             descr = task['description']
             exe   = descr['executable']
-            args  = descr.get('arguments', [])
-            cmd   = '%s(%s)' % (exe, ','.join(args)) 
+            args  = descr.get('arguments', list())
+            pres  = descr.get('pre_exec',  list())
+            cmd   = '%s(%s)' % (exe, ','.join(args))
 
             self._prof.prof('task_get', comp=wid, uid=tid)
           # self._log.debug('get %s: %s', tid, cmd)
 
-
             try:
-                task['stdout'] = eval(cmd)
+                for pre in pres:
+                    if pre.split()[0] == 'import':
+                        for mod in pre.split()[1:]:
+                            if mod not in globals():
+                                globals()[mod] = ru.import_module(mod)
+                                locals() [mod] = globals()[mod]
+
+                try:
+                    function_code = self.prepare_func(exe, args)
+                except Exception:
+                    self._log.debug("Setting the function for execution failed!.")
+                    raise
+                
+                @contextlib.contextmanager
+                def stdoutIO(stdout=None):
+                    old = sys.stdout
+                    if stdout is None:
+                       stdout = StringIO()
+                    sys.stdout = stdout
+                    yield stdout
+                    sys.stdout = old
+                
+                with stdoutIO() as s:
+                     exec(function_code)
+        
+                #task['stdout'] = eval(cmd) # We need to find a way to keey this option!!!
+                task['stdout'] = s.getvalue()
                 task['stderr'] = None
```

Comment:
Why don't we capture stderr also?

Reply:
On the exception check below: we could still obtain any `stdout` we captured so far.  Same for `stderr`, and then appending the exception stack maybe?  Note that RU has some [utiliy method](https://github.com/radical-cybertools/radical.utils/blob/devel/src/radical/utils/debug.py#L177) to obtain the exception stack as string.